connexion/apis/abstract.py

```diff
@@ -262,18 +364,19 @@ def add_operation(self, method, path, swagger_operation, path_parameters):
             "pass_context_arg_name": self.pass_context_arg_name
         }
 
-        if self.spec_version < (3, 0, 0):
+        # TODO refactor into AbstractOperation.from_spec(Specification, method, path)
+        if self.specification.version < (3, 0, 0):
             operation = Swagger2Operation(self,
-                                          app_produces=self.produces,
-                                          app_consumes=self.consumes,
-                                          security_definitions=self.security_definitions,
-                                          definitions=self.definitions,
-                                          parameter_definitions=self.parameter_definitions,
-                                          response_definitions=self.response_definitions,
+                                          app_produces=self.specification.produces,
+                                          app_consumes=self.specification.consumes,
+                                          security_definitions=self.specification.security_definitions,
+                                          definitions=self.specification.definitions,
+                                          parameter_definitions=self.specification.parameter_definitions,
+                                          response_definitions=self.specification.response_definitions,
                                           **shared_args)
         else:
             operation = OpenAPIOperation(self,
-                                         components=self.components,
+                                         components=self.specification.components,
                                          **shared_args)
```

Comment:
Can't we just pass the specification object as a whole?

Reply:
yeah, or see my TODO comment above.